### ISSUE TITLE
Folia support

### ISSUE THREAD

**Issue by MelonHell** [CitizensDev/Citizens2#3013]
### I confirm that I have ...

- [X] ...tried the latest [Development Build](https://ci.citizensnpcs.co/job/Citizens2/) to make sure my feature doesn't exist already.
- [X] ...checked the Wiki for any possible information about my feature.
- [X] ...checked other existing issues for my feature request.

### Description

https://github.com/PaperMC/Folia

**Comment by mcmonkey4eva** [comment]
As Paper's own team / Folia's developers will tell you: the majority of servers should *not* switch to Folia. Unless you *currently have* a 16+ core CPU and plural hundreds of concurrent players on the average day, you don't need nor want Folia.
Are you running a server that fits those very high requirements?

**Comment by MelonHell** [comment]
> As Paper's own team / Folia's developers will tell you: the majority of servers should *not* switch to Folia. Unless you *currently have* a 16+ core CPU and plural hundreds of concurrent players on the average day, you don't need nor want Folia.
> In fact, running below that level (too few cores or too few players), it's expected that Folia will run *worse* than a normal Paper server, due to the complex thread handling being done.
> Are you running a server that fits those very high requirements?

We personally expect 500-1000 online in one world and consider between Folia and Multipaper

In any case, I'm pretty sure that if a server-side software exists, then it should be used, and if it's used, then it should have core plugins like Citizens on it.  I understand that adding support for Folia is not easy, but on the other hand, this is not Sponge or Minestom, this is software that uses the Bukkit API and this is not a noname project that will be abandoned, the PaperMC team is trustworthy and I am sure this software will be distributed in the future and ported  Citizens on it is a justified investment of time and effort

**Comment by Aitooor** [comment]
Will the support be put on Folia?

**Comment by mani1232** [comment]
@Aitooor @MelonHell  try: https://github.com/FancyMcPlugins/FancyNpcs

**Comment by Aitooor** [comment]
> @Aitooor @MelonHell  try: https://github.com/FancyMcPlugins/FancyNpcs

That only work in the latest versions..

**Comment by DogLoverPink** [comment]
> > @Aitooor @MelonHell  try: https://github.com/FancyMcPlugins/FancyNpcs
> 
> That only work in the latest versions..

I believe that Folia is planned to only work on the more recent versions, as the Paper devs (reasonably so) don't like supporting legacy software.

**Comment by Aitooor** [comment]
> > > @Aitooor @MelonHell  try: https://github.com/FancyMcPlugins/FancyNpcs
> 
> > 
> 
> > That only work in the latest versions..
> 
> 
> 
> I believe that Folia is planned to only work on the more recent versions, as the Paper devs (reasonably so) don't like supporting legacy software.

Sorry, I explained it badly, what I mean is that if you enter with any other version of the server as a client, you will see an armorstand flying.

**Comment by mani1232** [comment]
> > > > @Aitooor @MelonHell  try: https://github.com/FancyMcPlugins/FancyNpcs
> > 
> > > 
> > 
> > > That only work in the latest versions..
> > 
> > 
> > 
> > I believe that Folia is planned to only work on the more recent versions, as the Paper devs (reasonably so) don't like supporting legacy software.
> 
> Sorry, I explained it badly, what I mean is that if you enter with any other version of the server as a client, you will see an armorstand flying.

Try znpcplus fork

**Comment by mani1232** [comment]
https://github.com/Pyrbu/ZNPCsPlus

**Comment by Aitooor** [comment]
> > > > > @Aitooor @MelonHell  try: https://github.com/FancyMcPlugins/FancyNpcs
> > > 
> > > > 
> > > 
> > > > That only work in the latest versions..
> > > 
> > > 
> > > 
> > > I believe that Folia is planned to only work on the more recent versions, as the Paper devs (reasonably so) don't like supporting legacy software.
> > 
> > Sorry, I explained it badly, what I mean is that if you enter with any other version of the server as a client, you will see an armorstand flying.
> 
> Try znpcplus fork

Sure, i already use that plugin branch 2.0.0

**Comment by MightyGoldenOctopus** [comment]
Hey it would be a great helps, if we can decouple the tick impact of NPCs across several tick loops it would allow us to scale Citizens far more easily, I realize it represents an entire project, a lot of organizations including mine would be ready to sponsor such an endeavour

**Comment by HydrolienF** [comment]
I'm running a 200+ online players server with Folia. And I would love to be able to use this plugin.

**Comment by RogerMinemu** [comment]
Hi. We're hosting a +400 players smp. Folia is our only to-go...

**Comment by fullwall** [comment]
I currently don't have the resources to dedicate 10s-100s of hours towards a rewrite to enable Folia support.

**Comment by ranminecraft** [comment]
any new?

**Comment by fullwall** [comment]
> any new?

Not at the moment. Read above.

**Comment by maxnovaone** [comment]
Def need a Folia support here. Thats the only way to go and yes we have enough hardware.
That singlethread schedule is not performant enough for large servers. The Test-Server of Folia is running unbelievable good with that multithreaded schedules. There are even something called universal-schedula or better MorePaperLib that runs on Bukkit and Folia and would be awesome to see that here in Citizens too. This is for sure the next thing for big servers. You can work with packets or the way to support all backwards mc versions. One PaperMC Dev told me that the way of packets will work and thats why we rewrite the most small plugins our self via sourcecode to bring them to work on Folia. But a Plugin like Citizens would be to high for us. Please add the Folia support. Thanks!

**Comment by fullwall** [comment]
> Def need a Folia support here. Thats the only way to go and yes we have enough hardware. That singlethread schedule is not performant enough for large servers. The Test-Server of Folia is running unbelievable good with that multithreaded schedules. There are even something called universal-schedula or better MorePaperLib that runs on Bukkit and Folia and would be awesome to see that here in Citizens too. This is for sure the next thing for big servers. You can work with packets or the way to support all backwards mc versions. One PaperMC Dev told me that the way of packets will work and thats why we rewrite the most small plugins our self via sourcecode to bring them to work on Folia. But a Plugin like Citizens would be to high for us. Please add the Folia support. Thanks!

Agreed. Rewriting Citizens is a big task, and I don't have the time at present (full time job). If you're a big server, then maybe you could consider donating resources in some way.

**Comment by SharkPika** [comment]
Could I inquire about the approximate progress of the project？My server urgently needs it！

**Comment by SharkPika** [comment]
> > Def need a Folia support here. Thats the only way to go and yes we have enough hardware. That singlethread schedule is not performant enough for large servers. The Test-Server of Folia is running unbelievable good with that multithreaded schedules. There are even something called universal-schedula or better MorePaperLib that runs on Bukkit and Folia and would be awesome to see that here in Citizens too. This is for sure the next thing for big servers. You can work with packets or the way to support all backwards mc versions. One PaperMC Dev told me that the way of packets will work and thats why we rewrite the most small plugins our self via sourcecode to bring them to work on Folia. But a Plugin like Citizens would be to high for us. Please add the Folia support. Thanks!
> 
> Agreed. Rewriting Citizens is a big task, and I don't have the time at present (full time job). If you're a big server, then maybe you could consider donating resources in some way.

How can I donate to obtain resources？My server urgently needs it！

**Comment by MightyGoldenOctopus** [comment]
Man this is open-source, not free on-demand software development, if you need it so much either become a sponsor and donate 50K USD a year or make a PR by yourself

**Comment by DavidDoesCode** [comment]
> As Paper's own team / Folia's developers will tell you: the majority of servers should _not_ switch to Folia. Unless you _currently have_ a 16+ core CPU and plural hundreds of concurrent players on the average day, you don't need nor want Folia. In fact, running below that level (too few cores or too few players), it's expected that Folia will run _worse_ than a normal Paper server, due to the complex thread handling being done. Are you running a server that fits those very high requirements?

This kind of thinking is dangerous - many servers are easily hitting their limit on player count and folia is one the better options of getting past this one-thread limit. This kind of thinking basically makes it only possible for "LARGE" servers to make use of folia - we should be working towards enabling servers of 30-100 players to make use of folia. (No papermc does not run 50-100 players with plugins or serious cuts to spawn rates)

**Comment by fullwall** [comment]
> This kind of thinking is dangerous - many servers are easily hitting their limit on player count and folia is one the better options of getting past this one-thread limit. This kind of thinking basically makes it only possible for "LARGE" servers to make use of folia - we should be working towards enabling servers of 30-100 players to make use of folia. (No papermc does not run 50-100 players with plugins or serious cuts to spawn rates)

It's not that we don't want to or can't work towards this, it's just that this is a very large effort to do. Ongoing feedback that this is a desired feature is helpful but really this is something that needs upwards of 50hours of developer time.

**Comment by Euphillya** [comment]
> It's not that we don't want to or can't work towards this, it's just that this is a very large effort to do. Ongoing feedback that this is a desired feature is helpful but really this is something that needs upwards of 50hours of developer time.

I can totally confirm! I'm working on it myself to make it compatible, and the work involved is enormous.

**Comment by DavidDoesCode** [comment]
>  > This kind of thinking is dangerous - many servers are easily hitting their limit on player count and folia is one the better options of getting past this one-thread limit. This kind of thinking basically makes it only possible for "LARGE" servers to make use of folia - we should be working towards enabling servers of 30-100 players to make use of folia. (No papermc does not run 50-100 players with plugins or serious cuts to spawn rates)
> 
> It's not that we don't want to or can't work towards this, it's just that this is a very large effort to do. Ongoing feedback that this is a desired feature is helpful but really this is something that needs upwards of 50hours of developer time.

100%. It's open source and anyone can do it themselves if they really want! 

I was just more so contributing to the "should it do be done?" debate.

**Comment by Euphillya** [comment]
Having real entities allows you to do a lot of things, like putting models into entities for example.
